server/routing/pubsub/manager.go

```diff
@@ -0,0 +1,282 @@
+	host        host.Host
+	pubsub      *pubsub.PubSub
+	topic       *pubsub.Topic
+	sub         *pubsub.Subscription
+	localPeerID string
+	topicName   string // Topic name (protocol constant)
+
+	// Callback invoked when label announcement is received
+	onLabelAnnouncement func(context.Context, *LabelAnnouncement)
+}
+
+// New creates a new GossipSub manager for label announcements.
+// This initializes the GossipSub router, joins the labels topic, and
+// starts the message handler goroutine.
+//
+// Protocol parameters (TopicLabels, MaxMessageSize) are defined in constants.go
+// and are intentionally NOT configurable to ensure network-wide compatibility.
+//
+// Parameters:
+//   - ctx: Context for lifecycle management
+//   - h: libp2p host for network operations
+//
+// Returns:
+//   - *Manager: Initialized manager ready for use
+//   - error: If GossipSub setup fails
+func New(ctx context.Context, h host.Host) (*Manager, error) {
+	// Create GossipSub with protocol-defined settings
+	ps, err := pubsub.NewGossipSub(
+		ctx,
+		h,
+		// Enable peer exchange for better peer discovery
+		pubsub.WithPeerExchange(true),
+		// Limit message size to protocol-defined maximum
+		pubsub.WithMaxMessageSize(MaxMessageSize),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create gossipsub: %w", err)
+	}
+
+	// Join the protocol-defined topic
+	topic, err := ps.Join(TopicLabels)
+	if err != nil {
+		return nil, fmt.Errorf("failed to join labels topic %q: %w", TopicLabels, err)
+	}
+
+	// Subscribe to receive label announcements
+	sub, err := topic.Subscribe()
+	if err != nil {
+		return nil, fmt.Errorf("failed to subscribe to labels topic %q: %w", TopicLabels, err)
+	}
+
+	manager := &Manager{
+		ctx:         ctx,
+		host:        h,
+		pubsub:      ps,
+		topic:       topic,
+		sub:         sub,
+		localPeerID: h.ID().String(),
+		topicName:   TopicLabels,
+	}
+
+	// Start message handler goroutine
+	go manager.handleMessages()
+
+	logger.Info("GossipSub manager initialized",
+		"topic", TopicLabels,
+		"maxMessageSize", MaxMessageSize,
+		"peerID", manager.localPeerID)
+
+	return manager, nil
+}
+
+// PublishLabels announces labels for a record to the network.
+// This is called when a record is stored locally and should be
+// discoverable by remote peers.
+//
+// Flow:
+//  1. Convert labels.Label to wire format ([]string)
+//  2. Create and validate LabelAnnouncement
+//  3. Publish to GossipSub topic
+//  4. GossipSub mesh propagates to all subscribed peers
+//
+// Parameters:
+//   - ctx: Context for operation timeout/cancellation
+//   - cid: Content ID of the record
+//   - labelList: List of labels extracted from the record
+//
+// Returns:
+//   - error: If validation or publishing fails
+//
+// Note: This is non-blocking. GossipSub handles propagation asynchronously.
+func (m *Manager) PublishLabels(ctx context.Context, cid string, labelList []labels.Label) error {
```

Comment:
perhaps a better way to do it is that we pass recordRef or a full record from which we construct the full publication event

```suggestion
func (m *Manager) PublishRecord(ctx context.Context, ref RecordRef) error {
```

Reply:
fixed